gdnative-core/src/geom/basis.rs

```diff
@@ -39,6 +106,201 @@ impl Basis {
+        let identity = Self::identity();
+        let m = (*self) * self.transposed();
+        m.is_equal_approx(&identity)
+    }
+
+    fn is_rotation(&self) -> bool {
+        let det = self.determinant();
+        det.approx_eq(&1.0) && self.is_orthogonal()
+    }
+
+    fn get_quat(&self) -> Quat {
+        debug_assert!(!self.is_rotation(), "Basis must be normalized in order to be casted to a Quaternion. Use get_rotation_quat() or call orthonormalized() instead.");
+
+        /* Allow getting a quaternion from an unnormalized transform */
+        let m = *self;
+        let trace = m.elements[0].x + m.elements[1].y + m.elements[2].z;
+        let mut temp = [0_f32; 4];
+
+        if trace > 0.0 {
+            let mut s = (trace + 1.0).sqrt();
+            temp[3] = s * 0.5;
+            s = 0.5 / s;
+
+            temp[0] = (m.elements[2].y - m.elements[1].z) * s;
+            temp[1] = (m.elements[0].z - m.elements[2].x) * s;
+            temp[2] = (m.elements[1].x - m.elements[0].y) * s;
+        } else {
+            let i = if m.elements[0].x < m.elements[1].y {
+                if m.elements[1].y < m.elements[2].z {
+                    2
+                } else {
+                    1
+                }
+            } else {
+                if m.elements[0].x < m.elements[2].z {
+                    2
+                } else {
+                    0
+                }
+            };
+
+            let j = (i + 1) % 3;
+            let k = (i + 2) % 3;
+
+            let elements_arr: Vec<[f32; 3]> = m.elements.iter().map(|e| e.to_array()).collect();
+
+            let mut s = (elements_arr[i][i] - elements_arr[j][j] - elements_arr[k][k] + 1.0).sqrt();
+            temp[i] = s * 0.5;
+            s = 0.5 / s;
+
+            temp[3] = (elements_arr[k][j] - elements_arr[j][k]) * s;
+            temp[j] = (elements_arr[j][i] + elements_arr[i][j]) * s;
+            temp[k] = (elements_arr[k][i] + elements_arr[i][k]) * s;
+        }
+
+        return Quat {
+            i: temp[0],
+            j: temp[1],
+            k: temp[2],
+            r: temp[3],
+            _unit: core::marker::PhantomData,
+        };
+    }
+
+    pub fn get_rotation_quat(&self) -> Quat {
+        // Assumes that the matrix can be decomposed into a proper rotation and scaling matrix as M = R.S,
+        // and returns the Euler angles corresponding to the rotation part, complementing get_scale().
+        // See the comment in get_scale() for further information.
+        let mut m = self.orthonormalized();
+        let det = m.determinant();
+        if det < 0.0 {
+            // Ensure that the determinant is 1, such that result is a proper rotation matrix which can be represented by Euler angles.
+            m.scale(&Vector3::new(-1.0, -1.0, -1.0));
+        }
+
+        return m.get_quat();
+    }
+
+    // Multiplies the matrix from left by the scaling matrix: M -> S.M
+    // See the comment for Basis::rotated for further explanation.
```

Comment:
This should be a doc-comment instead of a normal comment.

Reply:
Done.